fix issue #335

yichoi opened the pull request:
In order to avoid unnecessary copy, adding length check code

IoT.js-DCO-1.0-Signed-off-by: Youngil Choi duddlf.choi@samsung.com

LaszloLango commented:
Could you also add a regression test for the fix?

LaszloLango commented on src/iotjs_module_buffer.cpp:
```diff
@@ -216,7 +216,9 @@ JHANDLER_FUNCTION(Write) {
   BufferWrap* buffer_wrap = BufferWrap::FromJBufferBuiltin(*jbuiltin);
 
-  size_t copied = buffer_wrap->Copy(src.data(), 0, length, offset);
+  int length_min = src.size() < length ? src.size() : length;
```
What is the behavior of Node.js in such cases? Wouldn't be better to throw an exception if length parameter is bigger than the length of the source?

yichoi commented:
Since it is possible for users to use Buffer.prototype._builtin, I think every boundary checks (which currently exist only in js part) should also exist in native part, too. To avoid duplication, I just moved the range check code into native part.
And after comparing the current code with Node.js implementation like @LaszloLango said, I found that current implementation doesn't throw proper exception on some invalid-ranged indexes, so I added some more RangeErrors. (But I think it is unnecessary to throw RangeError in native code, because IMHO it is implementation-specific decision.)

BTW I think maybe we should re-consider about Node.js compatibility.

LaszloLango commented on test/run_pass/test_buffer_builtin.js:
```diff
@@ -0,0 +1,107 @@
+/* Copyright 2015 Samsung Electronics Co., Ltd.
```
Wrong copyright year

LaszloLango commented on src/iotjs_module_buffer.cpp:
```diff
@@ -82,6 +82,14 @@ size_t BufferWrap::length() {
+static size_t boundRange(int index, size_t low, size_t upper)
+{
+  if (index < static_cast<int>(low)) return low;
+  if (index > static_cast<int>(upper)) return upper;
```
I'm not sure what is the correct style, but this function looks odd to me. I think braces must be used in if statements and the left brace at the start of the function mustn't be on a separate line.

LaszloLango commented on test/run_pass/test_buffer_builtin.js:
```diff
@@ -0,0 +1,107 @@
+assert.equal(buff9._builtin.slice(0, -2).toString(), 'abcabcab');
+
+
+console.log(buff3.toString());
```
leftover debug code?

yichoi replied:
I copied that code from test_buffer.js, and I thought this code can be meaningful since it can check the result of toString (buf1) and slice (buf2).

LaszloLango replied:
AFAIK, tools/build.py redirects the output of the tests and do not print anything into the console, so I think it is useless to write `console.log` in test files. If you'd like to validate `buff3.toString()`, then I suggest to change `console.log` to `assert.equal`.

yichoi replied:
But https://github.com/Samsung/iotjs/blob/master/tools/check_test.py#L92 checks if @STDOUT exists in here: https://github.com/Samsung/iotjs/pull/393/files#diff-0f42bba455ef50958f24c64c145bfb91R17. It compares its console.log output, so if I remove the console.log, the test fails.

LaszloLango replied:
I missed that part. OK then.

LaszloLango commented on src/iotjs_module_buffer.cpp:
```diff
@@ -82,6 +82,17 @@ size_t BufferWrap::length() {
 }
 
 
+static size_t boundRange(int index, size_t low, size_t upper) {
```
`boundRange` -> `BoundRange`

LaszloLango commented on src/iotjs_module_buffer.cpp:
```diff
@@ -216,6 +235,10 @@ JHANDLER_FUNCTION(Write) {
 
   BufferWrap* buffer_wrap = BufferWrap::FromJBufferBuiltin(*jbuiltin);
 
+  offset = boundRange(offset, 0, buffer_wrap->length());
```
Isn't `BoundRange(offset, 0, buffer_wrap->length() - 1);` the correct call of this function?

yichoi replied:
If `write` function is called from buffer.js, out-of-range offset will throw RangeError, so it's OK because this line cannot be called.

And if `write` function is called directly from _builtin.write, I think it is more reasonable not to write any bytes in the buffer, if input offset was out-of-range. 
If we use `buffer_wrap->length() - 1`, `buffer_wrap->Copy()` will write 1 byte into buffer even if the offset was invalid.

LaszloLango replied:
The question was for every call of BoundRange. I'm worrying about the following corner case. If we have an example string "abcd" and call BoundRange like this

`BoundRande(4, 0, my_string->length());`

This will return 4, but my_String[4] is out of range. Did I missed something?

yichoi replied:
Yes, my_String[4] is out of range. But if I understand your concern correctly, it is *intended* out-of-range.
For example, if we have code like below,
```
var buff = new Buffer("abcd");
buff._builtin.write("ef", 100, 2);
```
The offset `100` is invalid, and nothing should be written in `buff`.
It is accomplished by passing out-of-range offset `4` into below copy function. (we won't enter into the for loop - https://github.com/Samsung/iotjs/blob/master/src/iotjs_module_buffer.cpp#L111)

I think every use of `BoundRange` have this kind of boundary condition.

LaszloLango commented:
LGTM

akosthekiss commented on src/iotjs_module_buffer.cpp:
```diff
@@ -82,6 +82,17 @@ size_t BufferWrap::length() {
 }
 
 
+static size_t BoundRange(int index, size_t low, size_t upper) {
```
The naming is a bit out-of-style: are we not starting function and method names with lower case letters, reserving upper case first letters for classes? (this would be `boundRange` then)

LaszloLango replied:
No. Grep for 'static void'  and you can see this is the correct style :) Did you find any example of the lowercase style? Is it mixed in the actual code base?

akosthekiss replied:
You are right. I must have mixed up something. My bad, sorry.

akosthekiss commented:
except for the style issue, the main content looks good

akosthekiss commented:
LGTM
